### PR TITLE
openssh: 8.1p1 -> 8.2p1

### PR DESCRIPTION
**Pull request by prusnak**
###### Motivation for this change

* https://www.openssh.com/txt/release-8.2
* add libfido2 to enable hardware tokens support added in this release

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by jonringer — COMMENTED** [review]
please fix eval issues

**Comment by prusnak** [comment]
eval problems seem unrelated to the package, but maybe @mmahut can confirm?

**Comment by prusnak** [comment]
@GrahamcOfBorg build openssh

**Comment by prusnak** [comment]
I just realized that gssapi patchset does not apply and we need to wait until the package is updated the debian (where we fetch the gssapi patch from).

**Comment by edolstra** [comment]
Or just disable the gssapi patchset. I don't think that should be a blocker for updating ssh.

File: pkgs/tools/networking/openssh/default.nix [hunk 1]
```diff
@@ -2,8 +2,10 @@
 , etcDir ? null
 , hpnSupport ? false
```
Comment (edolstra) [inline pkgs/tools/networking/openssh/default.nix]: This flag can be removed.

**Comment by prusnak** [comment]
I added another change in 464890f which keeps version 8.1p1 for openssh_gssapi. I kept openssh_hpn for now, because it was fixed by version update in 025217a.

If there is a reason to drop the `hpnSupport` completely because no one uses it, I can amend my
changes, but it seems it's not really required as it builds for now and the upstream seems active too.

**Comment by mmahut** [comment]
@GrahamcOfBorg eval

**Comment by fatpat** [comment]
tested with fido2 and a titan security key, works like a charm :+1:

**Comment by prusnak** [comment]
Can anyone help with the eval failure? I don't see what's wrong here ...

**Comment by mmahut** [comment]
@prusnak can you please try to rebase on master?

**Comment by prusnak** [comment]
@GrahamcOfBorg eval

**Comment by prusnak** [comment]
@GrahamcOfBorg build libfido2 openssh

**Comment by prusnak** [comment]
@GrahamcOfBorg build libfido2 openssh openssh_hpn openssh_gssapi

**Comment by prusnak** [comment]
@GrahamcOfBorg build libfido2 openssh openssh_hpn openssh_gssapi

**Comment by prusnak** [comment]
I had to fix the libfido2 on darwin, but afterward, it works!

**Comment by artemist** [comment]
It looks like libfido2 needs IOKit if you haven't fixed it yet

**Comment by prusnak** [comment]
Debian team meanwhile updated the gssapi patch for 8.2p1 so I also bumped openssh_gssapi to this version.

**Comment by prusnak** [comment]
@GrahamcOfBorg build libfido2 openssh openssh_hpn openssh_gssapi

**Comment by prusnak** [comment]
@GrahamcOfBorg build openssh openssh_hpn openssh_gssapi

**Comment by prusnak** [comment]
Once https://github.com/NixOS/nixpkgs/pull/80909 is merged I'll drop the `optional stdenv.isLinux` from libfido2 feature as this will be available on darwin too.

**Comment by prusnak** [comment]
Since #80909 was just merged, I updated my PR to enable FIDO2 support for Darwin too (replaced `optional stdenv.isLinux` with `optional stdenv.isUnix`).

**Comment by dtzWill** [comment]
I'm unable to work out a suitable fix, but just so folks are aware:
IIRC openssh+libfido2 causes eval infinite loop on musl-based
configurations, by means of NetBSD tool trying to fetch its source
via our fetchcvs which has openssh injected.

I'm not sure how to best resolve this, I tried a few things but
ultimately ran out of time and rebuild-power so it's on my "revisit"
pile.

Maybe gitMinimal doesn't need fido-enabled openssh?
* a144c0ef19ba51e24e436917683503c819e69449
(this helps via different route, maybe?)

Since netbsd CVS source doesn't need openssh AFAIK,
openssh-less fetchcvs and use it for these:

* 86c261db6d1eb2e7dd1471aa90e0ba6bdee45039

Or look at the problem and come up with a more coherent solution :).

Hope this helps, sorry can't tackle this myself for the next weeks :/.

**Comment by prusnak** [comment]
@GrahamcOfBorg build openssh openssh_hpn openssh_gssapi

**Comment by artemist** [comment]
This PR does need to add back a withFido2 option, which I think was in an earlier commit. I can't find it, but it should be in @prusnak's reflog. The option should probably default to `stdenv.isUnix`

**Comment by mmahut** [comment]
@GrahamcOfBorg test openssh

**Comment by prusnak** [comment]
I reintroduced the `withFIDO` option like suggested above by @artemist. However, there was no `stdenv.isUnix` attribute on my NixOS Linux system, so I used `stdenv.hostPlatform.isUnix` instead. Is that correct?

**Comment by mmahut** [comment]
@GrahamcOfBorg test openssh

**Review by mmahut — APPROVED** [review]
Successfully tested using my Trezor device [following this guide](https://blog.trezor.io/openssh-with-fido2-and-trezor-e565c2277).

**Comment by mmahut** [comment]
Given we need openssh 8.2 on the server side as well for these new features to work. I would like to propose to backport this into 20.03.

**Comment by prusnak** [comment]
@mmahut should we just create a PR to 20.03 branch or what exactly is the process for proposing a backport?

**Comment by worldofpeace** [comment]
We discussed the possibility of backporting openssh 8.2p1 in #nixos-dev 
* https://logs.nix.samueldr.com/nixos-dev/2020-02-27#3113587;

my feelings were well expressed in this comment https://logs.nix.samueldr.com/nixos-dev/2020-02-27#3113665;

A question that was raised that would make the determination easier if answered was "how long will 8.1 get security updates?".

**Comment by prusnak** [comment]
@worldofpeace I'll ask among the OpenSSH folks what's their security update strategy is.

**Comment by prusnak** [comment]
@worldofpeace answer from Damien Miller of OpenSSH: "We don't do backported security fixes - we're a tiny team and so only maintain the current release."

**Comment by worldofpeace** [comment]
> @worldofpeace answer from Damien Miller of OpenSSH: "We don't do backported security fixes - we're a tiny team and so only maintain the current release."

I guess this makes the decision easy :smile:

**Comment by worldofpeace** [comment]
@prusnak Let's do it.

**Comment by prusnak** [comment]
@worldofpeace Shall I prepare a PR? To which branch?

**Comment by worldofpeace** [comment]
> @worldofpeace Shall I prepare a PR? To which branch?

release-20.03

**Comment by prusnak** [comment]
Backport created in https://github.com/NixOS/nixpkgs/pull/81368

File: pkgs/tools/networking/openssh/default.nix [hunk 2]
```diff
@@ -41,15 +43,7 @@ stdenv.mkDerivation rec {
       ./dont_create_privsep_path.patch
 
       ./ssh-keysign.patch
```
Comment (flokli) [inline pkgs/tools/networking/openssh/default.nix]: what happened to `/detect_apple_ld.patch` from https://github.com/NixOS/nixpkgs/pull/81368 here?

Comment (prusnak) [reply pkgs/tools/networking/openssh/default.nix]: This patch was part of libfido2 package (added in other PR), not OpenSSH (this PR).

Comment (flokli) [reply pkgs/tools/networking/openssh/default.nix]: Okay, thanks!

**Pull request merged**